### PR TITLE
automagically generate API interface with Clang.jl

### PR DESCRIPTION
**Pull request by vchuravy**
So instead of creating a generator that we have to maintain constantly and where we also would have to carry every single version of `vulkan.h`, this uses Clang.jl to produce generate the API bindings.
### Todo:
- [ ] https://github.com/ihnorton/Clang.jl/pull/156 e.g. Vulkan Unions ...grrrr...
- [x] https://github.com/ihnorton/Clang.jl/pull/153 e.g. correctly handling pointers
- [x] https://github.com/ihnorton/Clang.jl/issues/157 WChar
- [x] VK_API_VERSION
- [x] Extensions

### CODE REVIEW

**Comment by vchuravy** [comment]
Added benefit is that compile down went down drastically. In order to check this out you will need https://github.com/JuliaGPU/Clang.jl/tree/vulkan

**Comment by vchuravy** [comment]
@SimonDanisch could you take a look and see if this covers the API fully enough? Do you need the xlib etc.. extensions?

**Comment by SimonDanisch** [comment]
> Do you need the xlib etc.. extensions?

It seems so ;) I need at least these guys: `VK_KHR_XCB_SURFACE_EXTENSION_NAME`

**Comment by SimonDanisch** [comment]
And you have some of the function arguments typed, e.g. `vkEnumeratePhysicalDevices`

File: gen/api/vk_1.0.3.jl [hunk 1]
```diff
@@ -0,0 +1,635 @@
...
+
+function vkCreateInstance(pCreateInfo,pAllocator,pInstance)
+    ccall((:vkCreateInstance,libvulkan),VkResult,(Ptr{VkInstanceCreateInfo},Ptr{VkAllocationCallbacks},Ptr{VkInstance}),pCreateInfo,pAllocator,pInstance)
+end
+
+function vkDestroyInstance(instance::VkInstance,pAllocator)
+    ccall((:vkDestroyInstance,libvulkan),Void,(VkInstance,Ptr{VkAllocationCallbacks}),instance,pAllocator)
+end
+
+function vkEnumeratePhysicalDevices(instance::VkInstance,pPhysicalDeviceCount,pPhysicalDevices)
```
Comment (vchuravy) [inline gen/api/vk_1.0.3.jl]: @SimonDanisch 

you mean `instance::VkInstance` here?

**Comment by vchuravy** [comment]
Using the rewriters from CUDArt and finally understanding how the clang arguments line works.  I managed to get a smaller version that is more functional.

The only missing thing I am aware of are constant definitions of the from https://github.com/ihnorton/Clang.jl/issues/158

File: gen/api/vk_1.0.3.jl (hunk 1 above)
Comment (SimonDanisch) [inline gen/api/vk_1.0.3.jl]: yes!

**Comment by vchuravy** [comment]
@SimonDanisch could you give this a last review and merge it? Unions are still problematic, but that can be fixed later on.

**Comment by SimonDanisch** [comment]
Oh no, this has the `Array_N_X`, we should definitely replace that with `NTuples`. 
I should have a patch lying around to adapt Clang to it ;)

**Comment by vchuravy** [comment]
Send me in the direction of that patch or open a PR with Clang.jl?

**Comment by SimonDanisch** [comment]
Here you go:
https://github.com/ihnorton/Clang.jl/pull/160

**Pull request merged**